### PR TITLE
hw4-patterns submission

### PR DESCRIPTION
**Pull request by vladyslavpavlenko**
- implement the Chain of Responsibility pattern for fetchers;
- add new fetchers (bank.gov.ua, api.privatbank.ua);
- add `FetcherWithLogger` for logging;
- refactor handlers repository;
- refactor `/cmd/api`;
- move JSON utility functions to `/pkg`;
- update tests.

### CODE REVIEW

**Comment by seemsod1** [comment]
approved

File: pkg/json/json_test.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 	"net/http/httptest"
 	"testing"
 
-	"github.com/vladyslavpavlenko/genesis-api-project/internal/utils"
+	json2 "github.com/vladyslavpavlenko/genesis-api-project/pkg/json"
```
Comment (vladpistun) [inline pkg/json/json_test.go]: better give normal name like jsonhelper

**Comment by vladpistun** [comment]
approved

File: pkg/json/json_test.go (hunk 1 above)
Comment (vladyslavpavlenko) [reply pkg/json/json_test.go]: oops, reasonable. thanks!

**Pull request merged**